### PR TITLE
Nofitted no paste

### PR DESCRIPTION
**Pull request by igorpnk**
Good afternoon! 
I created a script that allows you to disable paste for Not Fitted components in the current variant.
I want to offer it as a temporary way out of the situation here:
https://bugcrunch.live.altium.com/Idea/7562

### CODE REVIEW

File: Scripts - PCB/README.md [hunk 1]
```diff
@@ -31,6 +31,7 @@ Add-on name & download link | Script description
 [ManualToAutoDesignator](https://altium-designer-addons.github.io/DownGit/#/home?url=https://github.com/Altium-Designer-addons/scripts-libraries/tree/master/Scripts+-+PCB/ManualToAutoDesignator) | This script will change manual designators to auto-positioned. Will operate on all or selected components.
 [MoveAPdesignators](https://altium-designer-addons.github.io/DownGit/#/home?url=https://github.com/Altium-Designer-addons/scripts-libraries/tree/master/Scripts+-+PCB/MoveAPdesignators) | This script allows user to modify distance between auto-positioned designators and component
 [MoveToLayer](https://altium-designer-addons.github.io/DownGit/#/home?url=https://github.com/Altium-Designer-addons/scripts-libraries/tree/master/Scripts+-+PCB/MoveToLayer) | This script moves selected copper tracks to selected signal layer, while maintaining connectivity with Vias that are automatically placed.
+[NoFittedNoPaste](https://altium-designer-addons.github.io/DownGit/#/home?url=https://github.com/Altium-Designer-addons/scripts-libraries/tree/master/Scripts+-+PCB/NoFittedNoPaste) | Creates a rule that disables solder paste for all NotFitted components for the current Variant.
```
Comment (justinmassiot) [inline Scripts - PCB/README.md:34]:
```suggestion
[NofittedNoPaste](https://altium-designer-addons.github.io/DownGit/#/home?url=https://github.com/Altium-Designer-addons/scripts-libraries/tree/master/Scripts+-+PCB/NofittedNoPaste) | Creates a rule that disables solder paste for all NotFitted components for the current Variant.
```

File: Scripts - PCB/NofittedNoPaste/NofittedNoPaste.pas [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      if (CompVariation.DM_VariationKind = eVariation_NotFitted) then
+      begin
+           filter := filter + '(Component = ''' + Component.DM_PhysicalDesignator + ''') OR ';
+           NoFitye := false;
+      end;
+   end;
+  filter := copy(filter,1,Length(filter) - 4) + ')';
+
+  RulePaste := PCBServer.PCBRuleFactory(eRule_PasteMaskExpansion);
+  RulePaste.Expansion := MMsToCoord(-100);
```
Comment (justinmassiot) [inline Scripts - PCB/NofittedNoPaste/NofittedNoPaste.pas:75]: Would be good to have a number that is **really** impossible to obtain in practice (-10 000?).

Comment (igorpnk) [reply Scripts - PCB/NofittedNoPaste/NofittedNoPaste.pas:75]: I have never made pads more than 100mm. It seemed to me enough. But more is possible.

Comment (igorpnk) [reply Scripts - PCB/NofittedNoPaste/NofittedNoPaste.pas:75]: I tested -10000. The result of converting such a large number is incorrect: 909.217mm

Comment (igorpnk) [reply Scripts - PCB/NofittedNoPaste/NofittedNoPaste.pas]: Empirically, set -5000. Tested.

Comment (igorpnk) [reply Scripts - PCB/NofittedNoPaste/NofittedNoPaste.pas]: Additional tests showed that -5000 also failed.  -2539.975mm maximum.

**Comment by PetrTos** [comment]
Hi guys,
try to do some benchmarks. Altium native importers used a similar approach in the past where a huge negative number was used in Pad mask and paste expansion values (not in rule but in the objects itself). It turned out that it prevents some optimizations of online DRC and as a result, the environment got much slower. Online DRC determines the affected area after any change and checks only objects in that area. But with such huge objects (it ignored that it is a negative value) all over the place, it has to check the entire board all the time. You are using it in the Rule so it may not introduce the effect but it would worth a check how online DRC behaves with such rule.
 
BTW PCB editor has working area 100" by 100"s so in some places the 100" is the maximum number.

**Comment by justinmassiot** [comment]
Hello @PetrTos,
Thanks for your input.

I must say the online DRC is turned off in all my Altium Designer installations for years now :satisfied:
But ok you may be right, the negative distance hack is not desirable. Do you have something to propose to replace it completely? I think there is no such thing to "close the pastemask" in the pad's options.

**Pull request merged**